### ISSUE TITLE
Changefeed's next() blocks forever after idling for a certain amount of time

### ISSUE THREAD

**Issue by aeneasr** [rethinkdb/rethinkdb-go#344]
First of all, sorry for this lengthy issue. The TL;DR of this is: I believe that 20 minutes of idle time after connecting to the changefeed, causes the `Next()` method do block forever.

**Update: The issue was KeepAlive not being set. Read more on that [here](https://github.com/dancannon/gorethink/issues/344#issuecomment-233914778)**

I am connecting to rethinkdb using [this code](https://github.com/ory-am/hydra/blob/master/config/backend_connections.go#L41-L63). There are several `Changes()` subroutines. I believe all of them get disconnected and block excecution after a certain amount of time. The watcher I am investigating here is [this one](https://github.com/ory-am/hydra/blob/master/internal/fosite_store_rethinkdb.go#L238-L271). I was able to reproduce this issue with every other watcher too, for example [this one](https://github.com/ory-am/hydra/blob/master/client/manager_rethinkdb.go#L120-L151).

To help with debugging, I added a `log.Debug()` [here](https://github.com/ory-am/hydra/blob/master/internal/fosite_store_rethinkdb.go#L249) on my local machine. After connecting to RethinkDB initially, I can see that everything is working just fine:
## Fresh connection, 1-2 minutes after startup: everything is ok

**Client side**

```
(10:18 am) $ hydra token client
9O9QIK1ROvmsfyA3PDxFO5vyi4a9H4SgF2if7t3DYHI.0sAG3JSeEeC_q7fj_VRT2k_BGwFbnb-am3FQ7CFc1wQ
```

**Server side**

```
time="2016-07-20T10:18:40+02:00" level=debug msg="Received update from RethinkDB Cluster in OAuth2 manager."
```

I can confirm that the program was notified correctly, by running a validation command:

**Client side**

```
(10:18 am) $ hydra token validate 9O9QIK1ROvmsfyA3PDxFO5vyi4a9H4SgF2if7t3DYHI.0sAG3JSeEeC_q7fj_VRT2k_BGwFbnb-am3FQ7CFc1wQ

{
        "sub": "10f92d9b-d684-4fc0-96f9-7b5575c24119",
...
```

In the logs, I see that the access token was validated successfully, too:

**Server side**

```
time="2016-07-20T10:19:47+02:00" level=info msg="Access granted" audience=10f92d9b-d684-4fc0-96f9-7b5575c24119 request=&{rn:hydra:clients create 10f92d9b-d684-4fc0-96f9-7b5575c24119 map[owner:]} scopes=[hydra.clients] subject=10f92d9b-d684-4fc0-96f9-7b5575c2411
9
```
## Idling for ~20 minutes, `changes.Next(&update)` seems to be inresponsive/blocking

After idling for about 20 minutes, running

```
(10:39 am) $ hydra token client
9O9QIK1ROvmsfyA3PDxFO5vyi4a9H4SgF2if7t3DYHI.0sAG3JSeEeC_q7fj_VRT2k_BGwFbnb-am3FQ7CFc1wQ
```

works fine and yields a new token. In the RethinkDB management console I can see that a new document is inserted successfully in the right table. This is also backed by the fact that there is no error when running this command.

This time however, this

```
level=debug msg="Received update from RethinkDB Cluster in OAuth2 manager."
```

**is missing in the server side logs**, implying that `for changes.Next(&update)` did not trigger. I can confirm this by trying to validate the token, which this time results in a "resource not found":

```
$ hydra token validate OVX-RUWIRH3HYbuBH05bljk2EpqrACVMJkDacUA0COo.lZN-Ty95eXGvN_rFYggAdS3sVY811Rvf8OWbmyfhs-Y

Could not validate token: Got error (500): {"request":"8a55e6e8-ab65-4287-9dd5-c620da2940b5","error":"Could not find the requested resource(s): The request could not be authorized","code":500}
```

The complete log of the server side is:

```
time="2016-07-20T10:18:40+02:00" level=debug msg="Received update from RethinkDB Cluster in OAuth2 manager."

>> Here, a "Received update from RethinkDB Cluster in OAuth2 manager." at 10:39 is missing! <<

>> This error is caused by the manager who can't find the token because he wasn't notified by the change feed <<
time="2016-07-20T10:39:16+02:00" level=info msg="Got error." error="Could not find the requested resource(s): The request could not be authorized" request_id=8a55e6e8-ab65-4287-9dd5-c620da2940b5 stack="errors.wrapper Could not find the requested resource(s): Th
e request could not be authorized\nD:/workspace/go/src/github.com/ory-am/hydra/herodot/error.go:44 (0x78bb29)\n\tToError: ...............

```

I believe this issue is connected with #194 and maybe #318 . I am using commit hash `61cfdef59cb67a9bf9ab2311bd5eb69e132b6bf4` - I take it from the commit history that there where no bug fixes regarding this.

**Comment by aeneasr** [comment]
Does this have to do something with `KeepAlivePeriod`? I don't believe that either `ReadTimeout` nor `WriteTimeout` are the right options, or am I mistaken?

**Comment by aeneasr** [comment]
As already pointed out, it seems like `KeepAlivePeriod` not being set is the culprit here. I've set it to 10 seconds now and I can successfully listen to changefeed even after idling for a longer period of time:

```
time="2016-07-20T10:49:14+02:00" level=debug msg="Received update from RethinkDB Cluster in OAuth2 manager."
time="2016-07-20T11:15:40+02:00" level=debug msg="Received update from RethinkDB Cluster in OAuth2 manager."
```

I think there are multiple things to take away from this:
1. `Next()` should be able to detect when a connection timed out and return false
2. `KeepAlivePeriod` default value is `0` which in turn says: "Never send out any keep alive message". Instead, it should be a reasonable default value, like `time.Minute` or `30 * time.Second`

**Comment by dancannon** [comment]
Thanks for putting so much work into reporting this issue, you have no idea how much it helps! This issue is always hard to test as it takes so much time to reproduce, for anybody who comes across this issue (until a fix has been released) I recommend doing as @arekkas says and set the `KeepAlivePeriod` value to a non-zero value.

Both of your suggestions make sense and I will look into implementing both changes, hopefully I will have something ready to be released some time next week.

**Comment by aeneasr** [comment]
Welcome! Thanks for this amazing library!

**Comment by dancannon** [comment]
Just pushed 00c9457a255ad5a97b0e72484b8b10825f0b9c94 to `develop` which should address your second suggestion however I am not sure the first suggestion is possible. 

The reason for this is that in the situation you mentioned the client connection does not actually timeout as the connection does not have any timeouts by default. If you set `ReadTimeout` when creating a connection the cursor would actually return false when the connection timed out. If no timeouts are set then the client connection will keep waiting for data from the server even though the server may have dropped the connection (I may be wrong about this but thats what I believe is going on 😄 ).

Looking back at how the old code worked I made some incorrect assumptions about how both the `net` package worked and how users would configure their session (in future I will try to make sure I choose sensible values by default). Sorry for anybody who was affected by this issue!

**Comment by dancannon** [comment]
Just released `v2.1.2`, if possible could you rerun your tests to make sure that this issue has now been fixed?

**Comment by aeneasr** [comment]
> Just released v2.1.2, if possible could you rerun your tests to make sure that this issue has now been fixed?

I can confirm that the new default value for `KeepAlive` is resolving this issue. Maybe there is information out there which helps to identify the best possible value für KeepAlive? Depending on the application cluster nodes, high KeepAlive frequencies might be something to worry about.

**Comment by dancannon** [comment]
I chose 30s as this is the same value used by the `net/http` package, I think for now I will keep it simple as if applications require a higher frequency then they are able to change the value as required using `ConnectOpt`. Thanks again for all the help solving this issue.
